server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java

```diff
@@ -111,56 +109,44 @@ public HighlightField highlight(FieldHighlightContext fieldContext) {
+                        "] doc of [" + context.index().getName() + "] index " +
+                        "has exceeded [" + maxAnalyzedOffset + "] - maximum allowed to be analyzed for highlighting. " +
+                        "This maximum can be set by changing the [" + IndexSettings.MAX_ANALYZED_OFFSET_SETTING.getKey() +
+                        "] index level setting. " + "For large texts, indexing with offsets or term vectors, and highlighting " +
+                        "with unified or fvh highlighter is recommended!");
+            }
 
-                try (TokenStream tokenStream = analyzer.tokenStream(fieldType.name(), text)) {
-                    if (!tokenStream.hasAttribute(CharTermAttribute.class) || !tokenStream.hasAttribute(OffsetAttribute.class)) {
-                        // can't perform highlighting if the stream has no terms (binary token stream) or no offsets
-                        continue;
-                    }
-                    TextFragment[] bestTextFragments = entry.getBestTextFragments(tokenStream, text, false, numberOfFragments);
-                    for (TextFragment bestTextFragment : bestTextFragments) {
-                        if (bestTextFragment != null && bestTextFragment.getScore() > 0) {
-                            fragsList.add(bestTextFragment);
-                        }
+            try (TokenStream tokenStream = analyzer.tokenStream(fieldType.name(), text)) {
+                if (!tokenStream.hasAttribute(CharTermAttribute.class) || !tokenStream.hasAttribute(OffsetAttribute.class)) {
+                    // can't perform highlighting if the stream has no terms (binary token stream) or no offsets
+                    continue;
+                }
+                TextFragment[] bestTextFragments = entry.getBestTextFragments(tokenStream, text, false, numberOfFragments);
+                for (TextFragment bestTextFragment : bestTextFragments) {
+                    if (bestTextFragment != null && bestTextFragment.getScore() > 0) {
+                        fragsList.add(bestTextFragment);
                     }
                 }
-            }
-        } catch (Exception e) {
-            if (ExceptionsHelper.unwrap(e, BytesRefHash.MaxBytesLengthExceededException.class) != null) {
-                // this can happen if for example a field is not_analyzed and ignore_above option is set.
-                // the field will be ignored when indexing but the huge term is still in the source and
-                // the plain highlighter will parse the source and try to analyze it.
-                return null;
-            } else {
-                throw new FetchPhaseExecutionException(fieldContext.shardTarget,
-                    "Failed to highlight field [" + fieldContext.fieldName + "]", e);
+            } catch (BytesRefHash.MaxBytesLengthExceededException e) {
+                // ignore and continue to the next value
```

Comment:
We need special treatment because `Highlighter#getBestTextFragments` is declared as throwing `InvalidTokenOffsetsException`, which just extends plain `Exception`.  It doesn't feel correct rethrowing as an IOException (because it isn't one!) so wrapping in an IllegalArgumentException seems like a happy compromise.  It will be caught and translated to a FetchPhaseExecutionException by FetchPhase.

++ to add the explanatory comment back

Reply:
Got it, thanks! I had missed that it was a checked exception.

Reply:
Got it, thanks! I had missed that it was a checked exception.